### PR TITLE
Fix file format checks to be exact and handle Delta Lake column mapping [databricks]

### LINKED ISSUE #9255: [BUG] Unable to read DeltaTable with columnMapping.mode = name
**Describe the bug**
We would like to apply Spark-Rapids (23.08) plugin in Spark 3.3.0 environment in order to accelerate execution of SparkSQL with DeltaTable (2.3.0). However, we just discovered that Spark Rapids is not able to obtain data from a DeltaTable with following TBLPROPERTIES

```
+-------------------------------+-----+
|key                            |value|
+-------------------------------+-----+
|delta.columnMapping.maxColumnId|137  |
|delta.columnMapping.mode       |name |
|delta.minReaderVersion         |2    |
|delta.minWriterVersion         |5    |
+-------------------------------+-----+
```

**Steps/Code to reproduce bug**
Please provide a list of steps or a code sample to reproduce the issue.
You can create a DeltaTable with commands

```
df = spark.read.format("").load("load any of csv of json data")
spark.sql("SET spark.databricks.delta.properties.defaults.columnMapping.mode = name")
spark.sql("SET spark.databricks.delta.properties.defaults.minReaderVersion = 2")
spark.sql("SET spark.databricks.delta.properties.defaults.minWriterVersion = 5")
df.write.format("delta").saveAsTable("demo4")
```

**Expected behavior**
Once you create the delta table, you can use the following command to reproduce the error.

```
spark.sql("SELECT * FROM demo4").show()
```

Then, you might obtain results such as 

```
+----+----+
| _c0| _c1|
+----+----+
|null|null|
|null|null|
|null|null|
|null|null|
|null|null|
|null|null|
|null|null|
|null|null|
|null|null|
|null|null|
|null|null|
|null|null|
|null|null|
|null|null|
|null|null|
|null|null|
|null|null|
|null|null|
|null|null|
|null|null|
+----+----+
only showing top 20 rows
```

**Environment details (please complete the following information)**
 - Environment location: [Standalone]
 - Spark configuration settings related to the issue:
spark.driver.cores 1
spark.driver.memory 2g
spark.driver.extraClassPath /home/spark-current/jars/rapids-4-spark.jar
spark.executor.extraClassPath /home/spark-current/jars/rapids-4-spark.jar
spark.executor.cores 1
spark.executor.memory 2g
spark.executor.resource.gpu.discoveryScript /opt/sparkRapidsPlugin/getGpusResources.sh
spark.executor.resource.gpu.amount 1
spark.task.cpus 1
spark.task.resource.gpu.amount 1
spark.rapids.memory.pinnedPool.size 2g
spark.rapids.sql.concurrentGpuTasks 1
spark.rapids.sql.enabled true
spark.plugins com.nvidia.spark.SQLPlugin
spark.serializer org.apache.spark.serializer.KryoSerializer
spark.local.dir /mnt

**Additional context**

### PR DESCRIPTION
**Pull request by jlowe**
Fixes #9255.

This fixes the following problems:
- Format checks were allowing derived classes of the CPU class to be considered supported
- DeltaParquetFileFormat semantics were being completely ignored (i.e.; column mapping and deletion vectors)

File format checks were updated to check for specific CPU classes so we don't accidentally think we know how to replace the semantics when a derived class that alters the behavior is used instead.

The interface for Delta Lake providers was updated to handle read interfaces, and ExternalSource was updated accordingly.

### CODE REVIEW

**Comment by jlowe** [comment]
build

**Review by revans2 — COMMENTED** [review]
Looks good just one question, and there are some test failures so I am not going to approve it yet.

File: delta-lake/delta-20x/src/main/scala/com/nvidia/spark/rapids/delta/delta20x/Delta20xProvider.scala [hunk 1]
```diff
@@ -41,4 +49,31 @@ object Delta20xProvider extends DeltaIOProvider {
           .disabledByDefault("Delta Lake update support is experimental")
     ).map(r => (r.getClassFor.asSubclass(classOf[RunnableCommand]), r)).toMap
   }
+
+  override def isSupportedFormat(format: Class[_ <: FileFormat]): Boolean = {
+    format == classOf[DeltaParquetFileFormat] || format == classOf[GpuDelta20xParquetFileFormat]
```
[the remark this answers is not in the document]
Comment (revans2) [reply delta-lake/delta-20x/src/main/scala/com/nvidia/spark/rapids/delta/delta20x/Delta20xProvider.scala:54]: Okay could you add some docs to the isSupportedFormat declaration about who could call this and what is expected.

[the remark this answers is not in the document]
Comment (jlowe) [reply delta-lake/delta-20x/src/main/scala/com/nvidia/spark/rapids/delta/delta20x/Delta20xProvider.scala:54]: I decided this API is inherently confusing and fragile given it's looking at non-CPU formats.  I ended up solving this via #9283 and will update this PR to be based on that.  I'll also update the isSupportedFormat calls to check for the explicit CPU class we're expecting.

**Comment by jlowe** [comment]
Putting this in draft until it's been upmerged with #9283.

**Comment by LIN-Yu-Ting** [comment]
@jlowe. Very interesting development. I saw that 
1. you modified GpuFileSourceScanExec.scala to handle DeltaParquetFileFormat with ExternalSource.scala. 
2. DeltaProvider inside ExternalSource, implemented by DeltaProviderImpl.scala will execute function createMultiFileReaderFactory.
3. Inside, you will run createMultiFileReaderFactory based on delta table version and finally all of them call GpuDeltaParquetFileFormat to return ReaderFactory.

At the end, you prepare metadata such as StructType for DeltaTable with columnMappingMode.  
```
    GpuParquetMultiFilePartitionReaderFactory(
      fileScan.conf,
      broadcastedConf,
      prepareSchema(fileScan.relation.dataSchema),
      prepareSchema(fileScan.requiredSchema),
      prepareSchema(fileScan.readPartitionSchema),
      pushedFilters,
      fileScan.rapidsConf,
      fileScan.allMetrics,
      fileScan.queryUsesInputFile,
      fileScan.alluxioPathsMap)
```

which is different from original return value inside GpuFileSourceScanExec.
```
      GpuParquetMultiFilePartitionReaderFactory(
        sqlConf,
        broadcastedHadoopConf,
        relation.dataSchema,
        requiredSchema,
        readPartitionSchema,
        pushedDownFilters.toArray,
        rapidsConf,
        allMetrics,
        queryUsesInputFile,
        alluxioPathReplacementMap)
```

**Comment by jlowe** [comment]
> you prepare metadata such as StructType for DeltaTable with columnMappingMode

Yes, this reflects the behavior implemented by DeltaParquetFileFormat, the format used by the CPU when reading Delta Lake tables.  See https://github.com/delta-io/delta/blob/master/spark/src/main/scala/org/apache/spark/sql/delta/DeltaParquetFileFormat.scala#L128-L136

**Comment by jlowe** [comment]
build

**Review by revans2 — APPROVED** [review]

**Pull request merged**